### ISSUE TITLE
Docs: Formalize recommendation against plugins calling to rules via use-at-your-own-risk

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [eslint/eslint#19169]
### Docs page(s)

I don't know where this should go. Maybe some kind of _"recommendations for plugin integrators"_ and/or _"troubleshooting"_ page within https://eslint.org/docs/latest/extend/plugins?

### What documentation issue do you want to solve?

As seen in #19134 & https://github.com/typescript-eslint/typescript-eslint/issues/10338, some community plugins such as `@typescript-eslint/eslint-plugin` and `eslint-plugin-unicorn` rely on implementation details of ESLint core rules. [Paraphrased `builtinRules` usage from typescript-eslint](https://github.com/typescript-eslint/typescript-eslint/blob/3f02687d964c1e7888e5db62256d56cd72c4f281/packages/eslint-plugin/src/util/getESLintCoreRule.ts#L2):

```ts
import { builtinRules } from 'eslint/use-at-your-own-risk';

const baseRule = builtinRules.get('no-unused-expressions')!;

export default createRule({
  // ...
  create(context, [options]) {
    const rules = baseRule.create(context);

    return {
      ExpressionStatement(node) {
        // ...
        rules.ExpressionStatement(node);
      },
    };
  },
});
```

I'm told the ESLint team explicitly recommends not doing this practice. There's a reason why it's called _"use-at-your-own-risk"_: these rules are implementation details and should not be relied on.

### What do you think is the correct solution?

It'd be lovely to have this noted in docs. Both for visibility and so the plugins can align on a set of preferred strategies. https://github.com/typescript-eslint/typescript-eslint/issues/6456 tracks finding a better way to do things in typescript-eslint.

### Participation

- [x] I am willing to submit a pull request for this change.

### Additional comments

We should note that the recommendation isn't as simple as _"just don't do it!"_. Many ["extension" rules](https://typescript-eslint.io/rules/#extension-rules) such as [`@typescript-eslint/class-methods-use-this`](https://typescript-eslint.io/rules/class-methods-use-this) rely on working the same as ESLint core rules _except_ for specific cases. Plugins at the moment are stuck between a rock, a hard place, and a pit:

* ESLint core won't take in all of the extension behaviors (many wouldn't make sense to propose upstream at all; some were proposed but not accepted)
* Wrapping the base rule is an internal API per this recommendation and not supported
* Copying the rule makes for a much more difficult implementation on the plugin's side

This issue is not to track resolving those difficulties and tradeoffs. I'm just noting it here because the docs page needs to capture the nuance. We in the plugins haven't had a formal recommendation on the _right_ strategy prior to https://github.com/typescript-eslint/typescript-eslint/issues/6456#issuecomment-2495526641 - just indications of what not to do.

**Comment by nzakas** [comment]
Just off the top of my head, maybe the place to put this warning is on this page?
https://eslint.org/docs/latest/extend/custom-rules

So if someone is thinking they'll extend an existing rule and go that page looking for help, the warning will be there?

> We should note that the recommendation isn't as simple as "just don't do it!". 

Yes it is. :smile: We've said this over and over again: rules are *not* part of the public API and should not be loaded at runtime by anything other than ESLint itself. They can change, be removed, etc. at any point in time. When you do so, you're voiding your warranty and anything goes. The only [reason we have `builtinRules`](https://github.com/eslint/rfcs/blob/main/designs/2021-package-exports/README.md#remaining-public-api) is because too many existing plugins would break immediately. (Note: they *will* all break when we eventually move all rules out of the core.)

If we're going to document anything, it needs to be stated in the strongest possible terms: DO NOT DO THIS. The only acceptable way to extend a core rule is to copy the file into your own project and then wrap it however you want. That does put more maintenance burden on plugin developers, but that's the point. If you want to do something that isn't officially supported, you need to take full responsibility for that implementation. 

As an aside: We have releases every two weeks, so you could just set a job to run every two weeks to sync things up. Nothing will change any more rapidly than that.

Further discussions around this:
https://github.com/eslint/rfcs/pull/80

**Comment by nzakas** [comment]
@eslint/eslint-team other thoughts welcome

**Comment by mdjermanovic** [comment]
> maybe the place to put this warning is on this page?
> https://eslint.org/docs/latest/extend/custom-rules
>
> it needs to be stated in the strongest possible terms: DO NOT DO THIS. The only acceptable way to extend a core rule is to copy the file into your own project and then wrap it however you want.

This sounds good to me 👍

**Comment by fasttime** [comment]
We have a discussion in progress, so not stale.
